### PR TITLE
Getting the Metal target to run, adding OSX as a Travis CI target.

### PR DESCRIPTION
**Pull request by jamiltron**
I was able to get the Metal target to both run, and added a Travis CI target to try to keep it steady.

This should fix issue #1048 and issue #1014.

Sorry about my formatting - I realized my editor runs `rustfmt` on save, but it doesn't look like this entire code base is formatted this way, so these commits appear larger than they are. Do you have any kind of editor settings I could use to reformat this if this is a problem?

I know its kind of not a good idea to fix two things in a single pull request, but since it seemed like the Metal target rotted due to not having CI, I thought it might be best to add them in together. That way if I have to change the Metal code its still ran under CI, but also so that my pull request doesn't immediately cause the build to fail, if that stat is important.

I hope this helps out, and I would love pointers, tips, criticisms on things I could improve.

Thank you!

### CODE REVIEW

File: src/backend/metal/src/lib.rs [hunk 1]
```diff
@@ -176,50 +186,83 @@ impl core::Device for Device {
...
+                                                   }, // DSV
+                                                   |_, _v| {
+                                                       // v.sampler.release();
+                                                   }, // sampler
+                                                   |_, _| {
+                                                       // fence
+                                                   },
+                                                   |_, _| {
+                                                       // raw mapping
+                                                   });
```
Comment (Bastacyclop) [inline src/backend/metal/src/lib.rs]: This rustfmt formatting seems awful to me xD

File: src/backend/metal/src/map.rs [hunk 2]
```diff
@@ -440,77 +479,78 @@ pub fn map_base_type(ty: MTLDataType) -> shade::BaseType {
...
 
 pub fn map_texture_usage(usage: Usage) -> (MTLResourceOptions, MTLStorageMode) {
     match usage {
         Usage::GpuOnly => (MTLResourceStorageModePrivate, MTLStorageMode::Private),
-
-        Usage::Const   => (MTLResourceStorageModePrivate, MTLStorageMode::Managed),
+        Usage::Immutable => (MTLResourceStorageModePrivate, MTLStorageMode::Managed),
         Usage::Dynamic => (MTLResourceCPUCacheModeDefaultCache, MTLStorageMode::Managed),
         Usage::CpuOnly(access) => (map_access(access), MTLStorageMode::Managed),
+        Usage::Persistent(access) => (map_access(access), MTLStorageMode::Managed),
```
Comment (Bastacyclop) [inline src/backend/metal/src/map.rs]: Is this what we should return here ? (I have no idea)

File: src/backend/metal/src/map.rs [hunk 3]
```diff
@@ -440,77 +479,78 @@ pub fn map_base_type(ty: MTLDataType) -> shade::BaseType {
...
+        _ => unreachable!(),
     }
 }
 
 pub fn map_texture_usage(usage: Usage) -> (MTLResourceOptions, MTLStorageMode) {
     match usage {
         Usage::GpuOnly => (MTLResourceStorageModePrivate, MTLStorageMode::Private),
-
-        Usage::Const   => (MTLResourceStorageModePrivate, MTLStorageMode::Managed),
+        Usage::Immutable => (MTLResourceStorageModePrivate, MTLStorageMode::Managed),
```
Comment (Bastacyclop) [inline src/backend/metal/src/map.rs]: I think this should be [`MTLStorageMode::Private`](https://developer.apple.com/reference/metal/mtlstoragemode/1515797-private). Though you didn't change anything here so `Managed` might be necessary to initialize the immutable memory.

File: src/backend/metal/src/map.rs (hunk 2 above)
Comment (fkaa) [inline src/backend/metal/src/map.rs]: IIRC it might be just copy-pasted and I haven't gotten around to changing it yet, but fixing it shouldn't be part of this PR though

**Comment by fkaa** [comment]
I don't think the rustfmt should be too much of a problem (maybe we should do it over the whole codebase every now and then?).

Looks good to me, merging! 👍 
@homu r+

**Comment by kvark** [comment]
Oh wow, this is awesome. Thanks @jamiltron !

File: src/backend/metal/src/lib.rs (hunk 1 above)
Comment (jamiltron) [inline src/backend/metal/src/lib.rs]: Yeah there's some choices in rustfmt that I am kind of grossed out by, but personally I kind of think just agreeing on one standard kind of makes sense, even if you have to make some compromises on taste :P

File: src/backend/metal/src/map.rs (hunk 3 above)
Comment (jamiltron) [inline src/backend/metal/src/map.rs]: Yeah I wanted to keep most of the pre-existing code as is. I think you're right though so I'll check the documentation later and if so I'll give it a spin. Thanks!

File: src/backend/metal/src/map.rs (hunk 2 above)
Comment (jamiltron) [inline src/backend/metal/src/map.rs]: Hmm, I guess this depends on the target (iOS vs OSX) unless I'm mistaken ? Right now gfx doesn't support iOS, right? In that case I don't think shared is the right storage choice, given this statement in the documentation:

> In macOS, the shared storage mode is only available for MTLBuffer objects, not MTLTexture objects.

But then also it looks like for iOS and tvOS we'll have to change the storage, as [MTLStorageMode::Managed](https://developer.apple.com/reference/metal/mtlstoragemode/1515488-managed) is not available:

>  In iOS and tvOS, the managed storage mode is not available.

I can investigate further unless you were already working on this @fkaa

Thanks for the code review!

**Comment by jamiltron** [comment]
No problem, happy to help! Thanks for the awesome project!

File: src/backend/metal/src/map.rs (hunk 2 above)
Comment (fkaa) [inline src/backend/metal/src/map.rs]: It probably won't run on iOS and I don't have any iOS system to test it on unfortunately (simulator doesn't work either).

I have a bunch of changes locally that I'm going to push today or tomorrow which will make the metal backend actually useable, so I don't think you'd be able to even debug it in its current shape. I haven't touched the storage modes there though, so feel free to pick it up as soon as I hurry up and push the changes! 👍
